### ISSUE TITLE
Better shipping rates suggestions

### ISSUE THREAD

**Issue by ecgan** [woocommerce/google-listings-and-ads#1395]
## User story

This is related to https://github.com/woocommerce/google-listings-and-ads/pull/1341, #1382, #1383.

I was looking into the shipping rates suggestions changes, and I noticed there are a bit of inconsistencies (or, a gap in expectations) in the suggestions data returned. I'm hoping that via this issue, we can have a standardized expectation that can work across the back-end API and front-end UI.

In my WooCommerce shipping zones settings, I have the following setup:

![image](https://user-images.githubusercontent.com/417342/161308103-f59997b0-9b62-45fb-8963-92d065a7be18.png)

The zone names indicate the setting inside the zones: 

- "flat 0" means flat rate with cost 0.
- "flat 25" means flat rate with cost 25.
- "free shipping 0" means free shipping with no minimum amount. 
- "free shipping 50" means free shipping with minimum order amount 50.

With the above shipping zones setup, one would think there would be 12 shipping rates suggestions.

This is what I got when I call the suggestions API:

![image](https://user-images.githubusercontent.com/417342/161308563-040cd895-d0dc-49cc-ac04-df62580c422a.png)

All the countries are there in the response, but there are only 9 suggestions.

With the current shipping rate suggestion data structure, and the current shipping rate UI, it can become complicated and confusing.

**I would like to suggest a "standardized" shipping rate data and UI structure.**

Refer to the following annotated screenshot:

![image](https://user-images.githubusercontent.com/417342/161310612-c6cbe06f-a28c-44ca-9b18-91e6abb5411f.png)
 
In short, the idea is to convert WC shipping zone rates into "flat_rate" shipping rates in GLA as much as possible, with "free_shipping" complementing the "flat_rate". Note that this means that the shipping method in WC may not match with the shipping rate method in GLA, e.g. "free shipping" in WC would be converted to "flat rate" with amount 0 in GLA.

Here's the proposed suggestions API response based on the above WC shipping zones settings, with actual response from the above screenshot, and my proposed expected response: (I use ❓ to indicate "inconsistency" / "unknown" questions.)

- Indonesia (flat 0)
    - Actual: `{country: "ID", method: "flat_rate", currency: "USD", rate: 0, options: {}}`
    - Expected: `{country: "ID", method: "flat_rate", currency: "USD", rate: 0, options: {}}`
    - 🟢  Match.

- Myanmar (flat 25)
	- Actual: `{country: "MM", method: "flat_rate", currency: "USD", rate: 25, options: {}}`
	- Expected: `{country: "MM", method: "flat_rate", currency: "USD", rate: 25, options: {}}`
	- 🟢  Match.

- Malaysia (free shipping 0)
	- Actual: `{country: "MY", method: "free_shipping", currency: "USD", rate: 0, options: {}}`
	- Expected: `{country: "MY", method: "flat_rate", currency: "USD", rate: 0, options: {}}`
	- 🔴  Not match - should convert it to flat_rate.

- Vietnam (free shipping 50)
	- Actual `{country: "VN", method: "free_shipping", currency: "USD", rate: 0, options: {free_shipping_threshold: 50} }`
	- Expected: Do not return this suggestion.
	- 🔴  Not match - Vietnam has free shipping with minimum order amount 50, but does not have a flat rate. We should not return this suggestion. Previously I mentioned this in https://github.com/woocommerce/google-listings-and-ads/pull/1363#issuecomment-1078779535 and https://github.com/woocommerce/google-listings-and-ads/pull/1367#issuecomment-1079087068.

- Philippines (flat 0, free shipping 0)
	- Actual: `{country: "PH", method: "free_shipping", currency: "USD", rate: 0, options: {}}`
		- ❓ Note that there are two WC shipping zone methods, but there are only one returned. Why is that?
	- Expected: `{country: "PH", method: "flat_rate", currency: "USD", rate: 0, options: {}}`
	- 🔴  Not match - should use / convert to flat_rate.

- Singapore (flat 0, free shipping 50)
	- Actual: `{country: "SG", method: "flat_rate", currency: "USD", rate: 0, options: {}}` and `{country: "SG", method: "free_shipping", currency: "USD", rate: 0, options: {free_shipping_threshold: 50} }`
		- ❓ Why do we have two suggestions here?
	- Expected (assuming we are returning cheapest shipping rate, hence discarding the "free shipping 50"): `{country: "SG", method: "flat_rate", currency: "USD", rate: 0, options: {}}`
	- 🔴  Not match - the actual response would not be able to work with the current UI, because when we have flat rate 0, it does not make sense to enter minimum order amount 50 in the UI. 

- Thailand (flat 25, free shipping 0)
	- Actual: `{country: "TH", method: "free_shipping", currency: "USD", rate: 0, options: {}}`
	- Expected (assuming we are returning cheapest shipping rate, hence discarding the "flat 25"): `{country: "TH", method: "flat_rate", currency: "USD", rate: 0, options: {}}`
	- 🔴  Not match - should convert it to flat_rate.

- Sri Lanka (flat 25, free shipping 50)
	- Actual: `{country: "LK", method: "flat_rate", currency: "USD", rate: 25, options: {}}` and `{country: "LK", method: "free_shipping", currency: "USD", rate: 0, options: {free_shipping_threshold: 50} }`
	- Expected: `{country: "LK", method: "flat_rate", currency: "USD", rate: 25, options: {}}` and `{country: "LK", method: "free_shipping", currency: "USD", rate: 0, options: {free_shipping_threshold: 50} }`
	- 🟢  Match.

With the above expected API response, we would be able to better display the data in the UI. It would be clearer and consistent that "Estimated shipping rates" card UI works with "flat_rate", and "Minimum order" card UI works with "free_shipping", and this consistency applies to loading and saving shipping rates too (not just shipping rates suggestions).

### Describe the solution you'd like 

To implement the above proposal in the shipping rates suggestions API.

### Describe alternatives you've considered 

To implement the above proposal in the front-end UI, by doing a conversion process after getting the suggestions response. However, this isn't really ideal.

## Acceptance criteria

The API response is as above proposal.

**Comment by ecgan** [comment]
@nima-karimi , could you look into the above proposal and let me know if it makes sense? 🤔  🙏 

I was also confused with some ❓ inconsistencies, not sure if you have answers for them? By implementing the proposal, it should address those inconsistencies.

**Comment by nima-karimi** [comment]
> Malaysia (free shipping 0)
> Actual: {country: "MY", method: "free_shipping", currency: "USD", rate: 0, options: {}}
> Expected: {country: "MY", method: "flat_rate", currency: "USD", rate: 0, options: {}}
> 🔴 Not match - should convert it to flat_rate.

This is added in milestone two because now each method is mapped to a different class in the code and might have different options. As I mentioned, we can modify the suggestions to return these as `flat_rate` but the main goal was to have it simple and more or less identical to their representation in WC.

> Vietnam (free shipping 50)
> Actual {country: "VN", method: "free_shipping", currency: "USD", rate: 0, options: {free_shipping_threshold: 50} }
> Expected: Do not return this suggestion.
> 🔴 Not match - Vietnam has free shipping with minimum order amount 50, but does not have a flat rate. We should not return this suggestion. Previously I mentioned this in 
> https://github.com/woocommerce/google-listings-and-ads/pull/1363#issuecomment-1078779535 and https://github.com/woocommerce/google-listings-and-ads/pull/1367#issuecomment-1079087068.

The reasoning for including this was that there might be real-life cases where the customer only ships free if the order amount is a predefined number and offers a pickup or any other shipping method that we don't support for other cases. However, it might not make sense to include this by itself even if the user has other unsupported methods set up in WC. 

Since we are guesstimating the rates here, whether to include this or not is more of a personal preference. It could be argued that if a user has this rule it means that they DO offer free shipping for orders over a certain amount, regardless of whether they have any other rules set up, so it might be preferable for them to have Google at least offer the free shipping even if it doesn't support their other methods.

> Philippines (flat 0, free shipping 0)
> Actual: {country: "PH", method: "free_shipping", currency: "USD", rate: 0, options: {}}
> ❓ Note that there are two WC shipping zone methods, but there are only one returned. Why is that?
> Expected: {country: "PH", method: "flat_rate", currency: "USD", rate: 0, options: {}}
>  🔴 Not match - should use / convert to flat_rate.

If a location has an unconditional free shipping method (without any min order requirements) we ignore all the other methods and just return free shipping for it. This also means that if the user has other "unsupported" shipping methods in there, we still only return the free shipping method.

The reason that it returns "free_shipping" instead of "flat_rate" is also the same reason as Malaysia. 

> Singapore (flat 0, free shipping 50)
> Actual: {country: "SG", method: "flat_rate", currency: "USD", rate: 0, options: {}} and {country: "SG", method: "free_shipping", currency: "USD", rate: 0, options: {free_shipping_threshold: 50} }
> ❓ Why do we have two suggestions here?
> Expected (assuming we are returning cheapest shipping rate, hence discarding the "free shipping 50"): {country: "SG", method: "flat_rate", currency: "USD", rate: 0, options: {}}
>  🔴 Not match - the actual response would not be able to work with the current UI, because when we have flat rate 0, it does not make sense to enter minimum order amount 50 in the UI.

Currently, flat_rate and free_shipping are considered different methods so we won't mix them together. However, this does sound like a valid case for them to be combined. I'll set this up in a PR for this issue.


> Thailand (flat 25, free shipping 0)
> Actual: {country: "TH", method: "free_shipping", currency: "USD", rate: 0, options: {}}
> Expected (assuming we are returning cheapest shipping rate, hence discarding the "flat 25"): {country: "TH", method: "flat_rate", currency: "USD", rate: 0, options: {}}
>  🔴 Not match - should convert it to flat_rate.

The reason that it returns "free_shipping" instead of "flat_rate" is also the same reason in Malaysia and Philippines. 



Thanks for laying out all the combinations in detail. Some of the cases you mentioned as discrepancies are actually intentionally introduced in Milestone 2 because it would match better to how Google and Woo represent these values. I agree that some of them might not make sense in the form of suggestions but they do make the mapping easier when we are building data objects for submitting to Google API and syncing them. So the general idea was to have a similar data structure to Google and WC.

@ecgan Please let me know what you think about the above. We can modify the suggestions API and process and render some of these differently if they can't be handled in the front-end easily.

**Comment by ecgan** [comment]
@nima-karimi , thanks for the detailed explanation. 🙂 

---

I would like to take a step back from shipping rate suggestion API, and talk about shipping rate API first (saving and loading shipping rates). It relates to the plan I have in changing the UI code (issue https://github.com/woocommerce/google-listings-and-ads/issues/1382), and relates to your point quoted below: (with my emphasis)

> Some of the cases you mentioned as discrepancies are actually intentionally introduced in Milestone 2 because **it would match better to how Google and Woo represent these values**. I agree that some of them might not make sense in the form of suggestions but they do make the mapping easier **when we are building data objects for submitting to Google API and syncing them**. So the general idea was to have a **similar data structure to Google and WC**.

The plan that I have for the UI changes is illustrated below:

![image](https://user-images.githubusercontent.com/417342/162262436-d44d6f2b-c638-40d4-9560-5494a4809360.png)

- Values in the "Estimated shipping rates" card would have a corresponding shipping rate object with "flat_rate" method.
- Values in the "Minimum order" card would have a corresponding shipping rate object with "free_shipping" method.

So from the above screenshot, we will have the following shipping rates array data - 3 "flat_rate" and 1 "free_shipping":

```js
[
	// From "Estimated shiping rates" card:
	{
		country: 'LK', // Sri Lanka.
		method: 'flat_rate',
		currency: 'USD',
		rate: 14.0,
		options: {},
	},
	{
		country: 'MM', // Myanmar (flat_rate method with 0 rate; not free_shipping method).
		method: 'flat_rate',
		currency: 'USD',
		rate: 0,
		options: {},
	},
	{
		country: 'MY', // Malaysia
		method: 'flat_rate',
		currency: 'USD',
		rate: 12.0,
		options: {},
	},
	// From "minimum order" card:
	{
		country: 'LK', // Sri Lanka.
		method: 'free_shipping',
		currency: 'USD',
		rate: 0,
		options: {
			free_shipping_threshold: 88.0,
		},
	},
]
```

So an important question I have here is: would this shipping rate data work with your points I quoted above? i.e. Would it work with the syncing to Google API?

As a side note, regarding "it would match better to how Google _and Woo_ represent these values" and "the general idea was to have a similar data structure to Google _and WC_", I gave some more thoughts about this, and I'm not sure if it is good or important to have a similar structure to WC. 🤔 I mean, we are primarily interacting with Google API, so the data structure should primarily match with Google API - the "destination". The "source" of the data could come from manual user input (the UI form), or it could come from WC shipping settings. If it comes from WC shipping settings, then we can have a converter to convert WC shipping settings to our desired data structure in GLA to work with Google API. So in other words, I'm thinking maybe the data structure in GLA doesn't need to be similar with WC? If we try to make it similar with WC, it might be holding us back? 🤔 Maybe I'm missing something, like to hear your thoughts.

---

The above "shipping rate data" plan that I have is the main reason why I came up with this better shipping rates suggestions issue. I would like the suggestions to fit into the "shipping rate data" plan. Currently, the shipping rates suggestions data does not match with the plan, and if we were to use the plan, we would need to convert it, either in the front-end UI or in the API.

---

Next, I would like to talk about the "Vietnam" case:

> > Vietnam (free shipping 50)
> > Actual {country: "VN", method: "free_shipping", currency: "USD", rate: 0, options: {free_shipping_threshold: 50} }
> > Expected: Do not return this suggestion.
> > 🔴 Not match - Vietnam has free shipping with minimum order amount 50, but does not have a flat rate. We should not return this suggestion. Previously I mentioned this in
> > [#1363 (comment)](https://github.com/woocommerce/google-listings-and-ads/pull/1363#issuecomment-1078779535) and [#1367 (comment)](https://github.com/woocommerce/google-listings-and-ads/pull/1367#issuecomment-1079087068).
>
> ...[redacted for brevity]...
>  
> Since we are guesstimating the rates here, whether to include this or not is more of a personal preference. It could be argued that if a user has this rule it means that they DO offer free shipping for orders over a certain amount, regardless of whether they have any other rules set up, so it might be preferable for them to have Google at least **offer the free shipping** even if it doesn't support their other methods.

I guess what you are saying is that Google side is able to accept `{country: "VN", method: "free_shipping", currency: "USD", rate: 0, options: {free_shipping_threshold: 50} }` to display a message like "Free shipping for spend over USD 50" on Google properties, _without needing to have a shipping rate value for spend below USD 50_? 🤔 

With the current UI design and logic (tagging @j111q for your attention), unfortunately we are not able to do that though, because we need to have a flat rate method first before we can specify a free shipping method with minimum amount threshold. The "minimum order" card depends on the flat rates in "estimated shipping rate" card.

To move forward in the short term, if we can't remove the Vietnam data in the suggestions API, then I can figure out a way to remove the Vietnam data in the front-end side. This means that Google side won't have this Vietnam data.

For the long term, we probably need to have a UI redesign on this, to make the Vietnam data work and show up in Google.

**Comment by nima-karimi** [comment]
> Google side is able to accept {country: "VN", method: "free_shipping", currency: "USD", rate: 0, options: {free_shipping_threshold: 50} } to display a message like "Free shipping for spend over USD 50" on Google properties, without needing to have a shipping rate value for spend below USD 50?

Yes, there are no constraints set in their API to prevent this setting.

> To move forward in the short term, if we can't remove the Vietnam data in the suggestions API, then I can figure out a way to remove the Vietnam data in the front-end side. 

It should be fairly straightforward to filter these on the backend too but what I meant to say was that there might be actual use cases for this. And if the user chooses the "automatic" option it will be synced to Google.

> This means that Google side won't have this Vietnam data.

Through the suggestions API no, but if they choose the "automatic" option this rate will be synced to Google. Unless we remove it from all the mappings.

**Comment by nima-karimi** [comment]
> So an important question I have here is: would this shipping rate data work with your points I quoted above? i.e. Would it work with the syncing to Google API?

Yes as long as the "free_shipping_threshold" is only set on the "free_shipping" methods, it will work fine.

> I'm not sure if it is good or important to have a similar structure to WC.

It's not important but it will help reduce the code complexity to have an interim data object when parsing a complex set of data. Currently, the only thing that is similar to the WC structure is that we set the free_shipping type for free shipping rates. Another reason for doing this is that it will be easier to separate these rates into different classes and validate their options, etc. We could make it more similar to the way Google represents the data. For instance, all shipping rates can have a minimum order value as far as Google is concerned. But I'm not sure if this helps the front-end or just makes things more complicated?

> The above "shipping rate data" plan that I have is the main reason why I came up with this better shipping rates suggestions issue. I would like the suggestions to fit into the "shipping rate data" plan. Currently, the shipping rates suggestions data does not match with the plan, and if we were to use the plan, we would need to convert it, either in the front-end UI or in the API.

What do you think about the following changes?

- Return all free shipping rates as `flat_rate`. There will be no `free_shipping` method in the suggestions. A `flat_rate` with a cost of 0 can be considered free shipping. This will essentially revert some of the changes in milestone two to make the suggestions similar to milestone one.
- Do not return any suggestions for a country that has only a free_shipping rate with a threshold and no other rates.

**Comment by ecgan** [comment]
> For instance, all shipping rates can have a minimum order value as far as Google is concerned. But I'm not sure if this helps the front-end or just makes things more complicated?

Ooh, interesting. I'm not sure either, my gut feel says it might make things more complicated in the UI for the users. I think we can put that idea aside. 😂 

Thanks for letting me know the reasonings. 🙂 

> What do you think about the following changes?
> 
> * Return all free shipping rates as `flat_rate`. There will be no `free_shipping` method in the suggestions. A `flat_rate` with a cost of 0 can be considered free shipping. This will essentially revert some of the changes in milestone two to make the suggestions similar to milestone one.

How about the "free_shipping_threshold"? 🤔 If there is no "free_shipping" method, I guess it will show up as "flat_rate" with "options.free_shipping_threshold" like the following?

```js
	{
		country: 'LK', // Sri Lanka.
		method: 'flat_rate', // flat_rate, not free_shipping here.
		currency: 'USD',
		rate: 0,  // rate of 0, "free shipping".
		options: {
			free_shipping_threshold: 88.0,  // minimum order amount.
		},
	},
```

If this is the case, then this is essentially the same as milestone one, and there isn't any UI change needed? The current existing UI would continue to work as-is, no? 🤔 My assumption / expectation is that suggestions data and shipping rates data would share the same data structure. 

Or are you suggesting the above applies only to suggestions API, and then on the front-end UI, when we interact with shipping rate API (loading and saving shipping rates), we would still need to have "flat_rate" and "free_shipping" methods, and make sure that "_the "free_shipping_threshold" is only set on the "free_shipping" methods_"? If this is the case, it means the suggestions and the shipping rate data doesn't share the same structure, and on the front-end side, we would need to convert the suggestions into shipping rate data structure. The conversion here should be pretty simple and I can handle it if it is too complicated on the API side.

> * Do not return any suggestions for a country that has only a free_shipping rate with a threshold and no other rates.

Yeah, this sounds good. 🙂 🙏

**Comment by nima-karimi** [comment]
> How about the "free_shipping_threshold"? 🤔 If there is no "free_shipping" method, I guess it will show up as "flat_rate" with "options.free_shipping_threshold" like the following?

I was thinking it would be meaningless to have a "free_shipping" option on a "flat_rate" method so maybe it would be best to rename it to "minimum_order_amount"?

If you think that would complicate things, we can still return them as `free_shipping_threshold`.

> If this is the case, then this is essentially the same as milestone one, and there isn't any UI change needed? The current existing UI would continue to work as-is, no?

I think so, yes. If the UI code doesn't check or return "free_shipping" methods already then no need for any changes there.

**Comment by ecgan** [comment]
> I was thinking it would be meaningless to have a "free_shipping" option on a "flat_rate" method so maybe it would be best to rename it to "minimum_order_amount"?

If it's just changing the name to make it more meaningful, yeah we can do that, I can change it easily in the UI code. 🙂 

> I think so, yes. If the UI code doesn't check or return "free_shipping" methods already then no need for any changes there.

Yeah, right now the UI code doesn't have anything related to "free_shipping" method yet.

So I guess on my side, I'll just wait for the changes from your end, and when you are done, I'll do the necessary changes (e.g. the renaming to "minimum_order_amount" above) and test things out.

**Comment by nima-karimi** [comment]
Closing since it's already implemented in #1422.
